optaplanner-core/src/test/java/org/optaplanner/core/impl/score/director/ScoreDirectorSemanticsTest.java

```diff
@@ -0,0 +1,177 @@
+        TestdataConstraintConfigurationSolution solution2 =
+                TestdataConstraintConfigurationSolution.generateSolution(1, 1);
+        TestdataConstraintConfiguration constraintConfiguration = solution2.getConstraintConfiguration();
+        constraintConfiguration.setFirstWeight(SimpleScore.of(2));
+        scoreDirector.setWorkingSolution(solution2);
+        SimpleScore score2 = scoreDirector.calculateScore();
+        assertThat(score2).isEqualTo(SimpleScore.of(2));
+
+        // Set new solution with a disabled constraint, calculate score.
+        constraintConfiguration.setFirstWeight(SimpleScore.ZERO);
+        scoreDirector.setWorkingSolution(solution2);
+        SimpleScore score3 = scoreDirector.calculateScore();
+        assertThat(score3).isEqualTo(SimpleScore.ZERO);
+
+    }
+
+    @EnumSource(ScoreDirectorType.class)
+    @ParameterizedTest
+    void mutableConstraintConfiguration(ScoreDirectorType scoreDirectorType) {
+        InnerScoreDirectorFactory<TestdataConstraintConfigurationSolution, SimpleScore> scoreDirectorFactory =
+                scoreDirectorType.buildScoreDirectorFactory(solutionDescriptor);
+
+        // Create score director, calculate score with a given constraint configuration.
+        TestdataConstraintConfigurationSolution solution =
+                TestdataConstraintConfigurationSolution.generateSolution(1, 1);
+        InnerScoreDirector<TestdataConstraintConfigurationSolution, SimpleScore> scoreDirector =
+                scoreDirectorFactory.buildScoreDirector(false, false);
+        scoreDirector.setWorkingSolution(solution);
+        SimpleScore score1 = scoreDirector.calculateScore();
+        assertThat(score1).isEqualTo(SimpleScore.of(1));
+
+        // Change constraint configuration on the current working solution.
+        TestdataConstraintConfiguration constraintConfiguration = solution.getConstraintConfiguration();
+        scoreDirector.beforeProblemPropertyChanged(constraintConfiguration);
+        constraintConfiguration.setFirstWeight(SimpleScore.of(2));
+        scoreDirector.afterProblemPropertyChanged(constraintConfiguration);
+        SimpleScore score2 = scoreDirector.calculateScore();
+        assertThat(score2).isEqualTo(SimpleScore.of(2));
+    }
+
+    enum ScoreDirectorType {
+
+        EASY(new ScoreDirectorFactoryConfig()
+                .withEasyScoreCalculatorClass(DummyEasyScoreCalculator.class)),
+        INCREMENTAL(new ScoreDirectorFactoryConfig()
+                .withIncrementalScoreCalculatorClass(DummyIncrementalScoreCalculator.class)),
+        CS_DROOLS(new ScoreDirectorFactoryConfig()
+                .withConstraintProviderClass(DummyConstraintProvider.class)
+                .withConstraintStreamImplType(ConstraintStreamImplType.DROOLS)),
+        CS_BAVET(new ScoreDirectorFactoryConfig()
+                .withConstraintProviderClass(DummyConstraintProvider.class)
+                .withConstraintStreamImplType(ConstraintStreamImplType.BAVET)),
```

Comment:
Ordering: CS before INCREMENTAL, following the ordering of ScoreDirectorFactoryConfig's fields (which is enforced in the XSD ordering).